File: labs/control-ui/api-shadow/edge_worker_selector.ts
/**
 * Edge Worker Endpoint Selector and Latency Rater
 */
export interface WorkerEndpoint {
  domain: string;
  cleanIp: string;
  port: number;
  latencyMs: number;
  packetLossPct: number;
}

export class EdgeWorkerSelector {
  private endpoints: WorkerEndpoint[] = [];

  score(ep: WorkerEndpoint): number {
    return ep.latencyMs + ep.packetLossPct * 150.0;
  }

  addOrUpdate(
    domain: string,
    cleanIp: string,
    port: number,
    latencyMs: number,
    loss: number,
  ): void {
    const found = this.endpoints.find((e) => e.domain === domain && e.cleanIp === cleanIp);
    if (found) {
      found.latencyMs = latencyMs;
      found.packetLossPct = loss;
      found.port = port;
    } else {
      this.endpoints.push({
        domain,
        cleanIp,
        port,
        latencyMs,
        packetLossPct: loss,
      });
    }
  }

  selectBest(): WorkerEndpoint | null {
    if (this.endpoints.length === 0) return null;
    return [...this.endpoints].sort((a, b) => this.score(a) - this.score(b))[0]!;
  }

  synthesizeVlessUri(ep: WorkerEndpoint, uuid: string, sni: string): string {
    return `vless://${uuid}@${ep.cleanIp}:${ep.port}?encryption=none&security=tls&sni=${sni}&type=ws&host=${ep.domain}&path=%2F#LumiNet-Worker`;
  }
}

File: labs/mobile/donor-archive/tunnel/EdgeWorkerSelector.kt
package com.luminet.android.tunnel

data class WorkerEndpoint(
    val domain: String,
    val cleanIp: String,
    val port: Int,
    var latencyMs: Long,
    var packetLossPct: Float
) {
    fun score(): Double = latencyMs.toDouble() + (packetLossPct.toDouble() * 150.0)
}

class EdgeWorkerSelector {
    private val endpoints = mutableListOf<WorkerEndpoint>()

    fun addOrUpdate(domain: String, cleanIp: String, port: Int, latencyMs: Long, loss: Float) {
        val existing = endpoints.find { it.domain == domain && it.cleanIp == cleanIp }
        if (existing != null) {
            existing.latencyMs = latencyMs
            existing.packetLossPct = loss
        } else {
            endpoints.add(WorkerEndpoint(domain, cleanIp, port, latencyMs, loss))
        }
    }

    fun selectBest(): WorkerEndpoint? {
        return endpoints.minByOrNull { it.score() }
    }

    fun synthesizeVlessUri(ep: WorkerEndpoint, uuid: String, sni: String): String {
        return "vless://${uuid}@${ep.cleanIp}:${ep.port}?encryption=none&security=tls&sni=${sni}&type=ws&host=${ep.domain}&path=%2F#LumiNet-Worker"
    }
}
